macro/vignettes/pfSI.Rmd

```diff
@@ -0,0 +1,385 @@
+as two draws.
+
+Another case is when a change of state can lead to one of multiple transitions.
+In a Markov process, we would enumerate possible outcome states so that it's clear
+the system can't go to more than one. Then schedule them as competing processes.
+It can be easier to implement this behavior as a multinomial draw followed
+by scheduling the chosen event.
+
+
+# Continuous Time Model
+
+A continuous-time algorithm is a good choice when there will be few events in
+each time step. In this case, most processes are slow relative to a day, so we
+suspect continuous-time is a good choice, even in R where a single time step can 
+be expensive.
+
+We can make a back-of-the-envelope calculation by assuming larger numbers for
+each possible event. If each person gets ten bouts of
+malaria in their lifetime, that's infection, treatment, and back to susceptible, for
+thirty events. If we consider infectious bites, it's either those same infection events
+or we're counting a lot more bites, in the thousands. For a ten year simulation, it's
+3600 days in ten years. We either have events in 30/3600 days or over one per day.
+In the former case, continuous time would likely do better. In the latter case, discrete
+time has an advantage.
+
+
+## Implementation
+
+We can represent the state above as a disease state, treatment state,
+and scheduled bites.
+```{r list_state}
+individual <- list(
+  disease = factor("S", levels = c("S", "I")),
+  treatment = factor("U", levels = c("U", "PE", "GS")),
+  bites = c(0.1, 0.4, 5.7, 7.4)
+  )
+```
+Then we have about twelve transitions. We usually think about these rules
+as, "when this happens, then that can happen," but we can, alternatively,
+think about each transition as a separate entity that can happen whenever
+the state allows it. In this case, each one is defined by
+
+- is_enabled, a rule to determine from the new state whether the
+  transition could happen.
+- when, a function that reads the current state to draw a next time to fire.
+  If the distribution is incomplete, this returns infinity for not firing.
+- fire, a function that modifies the state when this fires.
+
+This kind of definition decouples the transitions from each other.
+It gets rid of questions about disabling from the queue, at the expense
+of checking, at every time step, which is enabled.
+
+We could store these as named lists.
+```{r transition_list}
+program_globals <- new.env()
+program_globals$b <- 0.2
+program_globals$c <- 3.7
+
+is_enabled <- list(
+  infect = function(state) with(state, disease == "S"),
+  recover = function(state) with(state, disease == "I")
+)
+when <- list(
+  infect = function(state) rexp(1, 1/50),
+  recover = function(state) rexp(1, 1/200)
+)
+fire <- list(
+  infect = function(state, time) {
+    b <<- 0.3
+    within(state, {disease = "I"})
+    },
+  recover = function(state, time) {
+    b <<- 0.25
+    within(state, {disease = "S"})
+    }
+)
+for (set_env_idx in 1:length(fire)) {
```

Comment:
This line is gonna need some extra explanation to people that it's setting the environment for each function to look up symbol values to be the `program_globals` environment. It's slick.

Reply:
This is something I needed to know. I didn't guess environments are less often used.